Pixel cluster counting code from Run 1

capalmer85 opened the pull request:
This code is mainly for historical purposes.  It will be updated for Run 2 format shortly.  I have only attempted to use this code in 53X for the purpose of validation on Run 1 data.

capalmer85 committed:
Dan's pixel cluster counting code from Run 1

capalmer85 commented:
Hello.  Happy new year!  Can someone please merge or explain why I need to rebase?  Rebasing appears to not be necessary and the pull request should be automatically merge-able from the information listed here.

capalmer85 commented:
@smuzaffar Can you explain the situation with my pull request?  I am not certain why I should need to rebase.  If I do need to rebase, can you please tell me what I should rebase to?

cmsbuild commented:
+1
Tested at: dbfecb4acd186b2ed34e219afd2a225e78d36c0b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6962/2414/summary.html

slava77 commented:
Sorry for missing this.
If this code is for Lumi pog, please check with the conveners. 

@plujan is there a place for this in a /test directory for an existing recolumi package? 
We usually don't keep ntuplizers in plugins in cmssw.

capalmer85 commented:
Hi Slava,

Marco Zanetti is in cc.  He can confirm that this code is blessed by the
LUMI POG.

Cheers,
Chris

On Fri, Feb 6, 2015 at 11:28 AM, Slava Krutelyov <notifications@github.com>
wrote:

> Sorry for missing this.
> If this code is for Lumi pog, please check with the conveners.
>
> @plujan <https://github.com/plujan> is there a place for this in a /test
> directory for an existing recolumi package?
> We usually don't keep ntuplizers in plugins in cmssw.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/6962#issuecomment-73216656>.
>

plujan commented:
@slava77 No, it doesn't really fit in the existing packages in RecoLuminosity -- it doesn't use the Lumi DB, so it doesn't make sense to put it in LumiDB/, and it's not online, so it doesn't belong in LumiProducer/.

It could probably use a less generic name, though, maybe PixelLumiAnalyzer?

capalmer85 commented:
OK.  So the folder name should be changed and it should go somewhere else.
Are there some suggestions for where that should be?

On Fri, Feb 6, 2015 at 11:49 AM, plujan <notifications@github.com> wrote:

> @slava77 <https://github.com/slava77> No, it doesn't really fit in the
> existing packages in RecoLuminosity -- it doesn't use the Lumi DB, so it
> doesn't make sense to put it in LumiDB/, and it's not online, so it doesn't
> belong in LumiProducer/.
>
> It could probably use a less generic name, though, maybe PixelLumiAnalyzer?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/6962#issuecomment-73219066>.
>

slava77 commented:
Is this code going to be used regularly for prompt calibrations or similar things?

Or is it really just a once-per-PAS/Paper application, used primarily by just one analysis group?
For the latter, the more appropriate place is outside of CMSSW, in 
https://github.com/cms-analysis

slava77 commented:
For a more frequent use 
RecoLuminosity/LumiProducer/test/analysis
or
RecoLuminosity/LumiProducer/analysis
makes sense to me.

capalmer85 commented:
This code (well, the newer version of it) will be used on a regular basis.
They will be needed for lumi validation and calibration.  We were hoping to
get these ntuples produced centrally and so we wanted to put the code
somewhere "official".

On Fri, Feb 6, 2015 at 2:48 PM, Slava Krutelyov <notifications@github.com>
wrote:

> Is this code going to be used regularly for prompt calibrations or similar
> things?
>
> Or is it really just a once-per-PAS/Paper application, used primarily by
> just one analysis group?
> For the latter, the more appropriate place is outside of CMSSW, in
> https://github.com/cms-analysis
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/6962#issuecomment-73238356>.
>

slava77 commented:
@diguida @monttj please take a note and maybe comment


If this code will be used regularly, then the case for having it in cmssw is valid.

Instead of making a new package with just one source code file,
the first guess would be to find a best-fitting existing package.
Hence the suggestion for RecoLuminosity/LumiProducer/analysis

If this will develop into something bigger,
it's more practical to have it in a separate package
It doesn't seem like this code will be a part of the standard 
reconstruction sequence, so, a Reco* subsystem is inappropriate for a big thing.

Will you produce any output that goes into the database?
It's unclear to me where analyzers targeting LumiDB fit, it's not the same DB as AlCa.
For this we have similar analyzers for AlCa, these are usually sitting
in Calib* subsystem(s) ==>  CalibLuminosity/TrackerBasedCalibration

Another alternative, more analysis-like is *Analysis subsystem,
e.g., LumiAnalysis/TrackerAnalyzers

capalmer85 commented:
While this code is going to evolve, I do not expect it to grow
significantly in size or scope.

I'm not sure that we will put any information into the LumiDB directly from
these ntuples.  In Run 1, the process (as I have understood it) was to use
the PCC analysis to produce functions to correct the HF lumi.

I think that putting everything in RecoLuminosity/LumiProducer/analysis
should work for me.  To go forward with this plan should I make an
additional pull-request?

@plujan <https://github.com/plujan> Is this ok for you?  If you have a
further suggestion, please let me know.


On Fri, Feb 6, 2015 at 4:18 PM, Slava Krutelyov <notifications@github.com>
wrote:

> @diguida <https://github.com/diguida> @monttj <https://github.com/monttj>
> please take a note and maybe comment
>
> If this code will be used regularly, then the case for having it in cmssw
> is valid.
>
> Instead of making a new package with just one source code file,
> the first guess would be to find a best-fitting existing package.
> Hence the suggestion for RecoLuminosity/LumiProducer/analysis
>
> If this will develop into something bigger,
> it's more practical to have it in a separate package
> It doesn't seem like this code will be a part of the standard
> reconstruction sequence, so, a Reco* subsystem is inappropriate for a big
> thing.
>
> Will you produce any output that goes into the database?
> It's unclear to me where analyzers targeting LumiDB fit, it's not the same
> DB as AlCa.
> For this we have similar analyzers for AlCa, these are usually sitting
> in Calib* subsystem(s) ==> CalibLuminosity/TrackerBasedCalibration
>
> Another alternative, more analysis-like is *Analysis subsystem,
> e.g., LumiAnalysis/TrackerAnalyzers
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/6962#issuecomment-73251829>.
>

slava77 commented:
if you can stay on the same github branch, then no need to resubmit a  PR, changes will be picked up automatically.

slava77 commented:
I suggest to cleanup the code as well.
There seem to be a lot of useless includes and declared dependencies:
- Many *Muon*  headers, but nothing is read from the event related to muons
.. I'll post a few more inline

slava77 commented on RecoLuminosity/LumiAnalyzer/plugins/LumiAnalyzer.cc:
```diff
@@ -0,0 +1,416 @@
+#include "CondFormats/L1TObjects/interface/L1GtTriggerMenu.h"
+#include "CondFormats/L1TObjects/interface/L1GtTriggerMenuFwd.h"
+
+#include "DataFormats/L1GlobalMuonTrigger/interface/L1MuRegionalCand.h"
```
L1Mu isn't used either, looks like a spurious unneeded header.

slava77 commented on RecoLuminosity/LumiAnalyzer/plugins/LumiAnalyzer.cc:
```diff
@@ -0,0 +1,416 @@
+#include "DataFormats/L1GlobalTrigger/interface/L1GlobalTriggerObjectMapRecord.h"
+#include "DataFormats/HLTReco/interface/TriggerEventWithRefs.h"
+#include "DataFormats/HLTReco/interface/TriggerObject.h"
+#include "DataFormats/HLTReco/interface/TriggerFilterObjectWithRefs.h"
```
None of the *Trigger* includes seem to be used either

slava77 commented on RecoLuminosity/LumiAnalyzer/plugins/LumiAnalyzer.cc:
```diff
@@ -0,0 +1,416 @@
+#include "MagneticField/Engine/interface/MagneticField.h"
+#include "MagneticField/Records/interface/IdealMagneticFieldRecord.h"
+
+#include "RecoMuon/GlobalTrackingTools/interface/GlobalMuonTrackMatcher.h"
```
not used

slava77 commented on RecoLuminosity/LumiAnalyzer/plugins/LumiAnalyzer.cc:
```diff
@@ -0,0 +1,416 @@
+	  }
+	}
+      }
+    }
```
~up to 20 lines appear to compute something that's not needed

capalmer85 replied:
Hi Slava,

Thanks for the feedback.  Indeed, this code could use a little cleaning.  I
hadn't bothered to do much of that for this pull request because I intended
to follow up with a newer version which is also less cluttered.

I will implement the cleaning you've suggested in this request and
forthcoming pull requests will be cleaner.

I am about to go on vacation for a week.  If I find a moment to sort this
out I'll let you know.  Otherwise, I'll take care of it the week after next.

Cheers,
Chris



On Fri, Feb 6, 2015 at 9:24 PM, Slava Krutelyov <notifications@github.com>
wrote:

> In RecoLuminosity/LumiAnalyzer/plugins/LumiAnalyzer.cc
> <https://github.com/cms-sw/cmssw/pull/6962#discussion_r24269174>:
>
> > +    edm::ESHandle<TrackerGeometry> pDD;
> > +    iSetup.get<TrackerDigiGeometryRecord>().get(pDD);
> > +
> > +    for (TrackerGeometry::DetContainer::const_iterator it = pDD->dets().begin(); it != pDD->dets().end(); it++){
> > +      if(dynamic_cast<PixelGeomDetUnit*>((*it))!=0){
> > +	DetId detId = (*it)->geographicalId();
> > +	uint32_t newDetId = detId;
> > +	
> > +	for (int fedid = 0; fedid < 40; ++fedid) {
> > +	  if (fPFC[fedid]->hasDetUnit(newDetId)) {
> > +	    fFEDID.insert(make_pair(newDetId, fedid));
> > +	    break;
> > +	  }
> > +	}
> > +      }
> > +    }
>
> ~up to 20 lines appear to compute something that's not needed
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/6962/files#r24269174>.
>

monttj commented:
Hi,

If it is just one source code and it is not part of reconstruction chain,
you can also indeed consider using existing analysis package. 
PhysicsTools/RecoAlgos/
In this way, you would have more flexibility when the code evolves. 

Regards,
Taejeong

capalmer85 committed:
rename and cleaning up

capalmer85 commented:
I've cleaned up the code and renamed the directory to a meaningful name.  I've also attempted to put the code into RecoLuminosity/LumiProducer/analysis, but this didn't work very well for me.  I couldn't figure out how to not compile everything that is in LumiProducer already.  This code is not dependent on anything in LumiProducer so I'm not sure what technical motivation we have to place it there.

Please review.

cmsbuild commented:
Pull request #6962 was updated. @cmsbuild, @nclopezo can you please check and sign again.

capalmer85 commented:
Hi Slava,

Could you take a look at this when you have some time?  (If this is the
wrong Slava, sorry to bother you.)

Thanks,
Chris

On Tue, Feb 24, 2015 at 9:12 AM, cmsbuild <notifications@github.com> wrote:

> Pull request #6962 <https://github.com/cms-sw/cmssw/pull/6962> was
> updated. @cmsbuild <https://github.com/cmsbuild>, @nclopezo
> <https://github.com/nclopezo> can you please check and sign again.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/6962#issuecomment-75762302>.
>

slava77 commented:
Hi Chris,

Lets try to get it in the RecoLuminosity/LumiProducer/analysis

capalmer85 commented:
Ok.  I think I need some offline help with that.  I tried just mv the directory and compiling.  This compiled the LumiProducer but nothing in analysis.  I would like to be able to compile only this code and not the LumiProducer since there are no dependences.  How should I proceed?  Thanks.

capalmer85 committed:
first attempt to do LumiProducer integration...

capalmer85 committed:
fix build file and typo

capalmer85 commented:
Hi Slava,

I've integrated into RecoLuminosity/LumiProducer/test/analysis.  Please
take a look.

Cheers,
Chris

On Thu, Feb 26, 2015 at 11:06 AM, Slava Krutelyov <notifications@github.com>
wrote:

> Hi Chris,
>
> Lets try to get it in the RecoLuminosity/LumiProducer/analysis
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/6962#issuecomment-76205204>.
>

cmsbuild commented:
Pull request #6962 was updated. @cmsbuild, @cvuosalo, @nclopezo, @slava77 can you please check and sign again.

capalmer85 commented:
Hi Slava,

Does this look ok?

Cheers,
Chris

On Mon, Mar 2, 2015 at 3:08 AM, cmsbuild <notifications@github.com> wrote:

> Pull request #6962 <https://github.com/cms-sw/cmssw/pull/6962> was
> updated. @cmsbuild <https://github.com/cmsbuild>, @cvuosalo
> <https://github.com/cvuosalo>, @nclopezo <https://github.com/nclopezo>,
> @slava77 <https://github.com/slava77> can you please check and sign again.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/6962#issuecomment-76672009>.
>

slava77 commented:
Sorry, I didn't get around to check this.
The next 53X release deadline in Mar 17.
We should get it done by then.

capalmer85 commented:
This is the old, historical version of this code.  I'd like to check in the
newer version of this (which is basically done already) before March 17th
and then start working on the 74X version of this code.

Can we finish with this in the next couple of days so I can proceed with
the run2 development soon afterward please?

Cheers,
Chris

On Tue, Mar 3, 2015 at 11:14 AM, Slava Krutelyov <notifications@github.com>
wrote:

> Sorry, I didn't get around to check this.
> The next 53X release deadline in Mar 17.
> We should get it done by then.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/6962#issuecomment-76976388>.
>

slava77 commented:
@cmsbuild please test

[I'm not sure though if jenkins tests are setup to run on 53X]

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 53f89817a721650a85324a7279c5ca3ce7d3db19
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6962/3169/summary.html

slava77 commented on RecoLuminosity/LumiProducer/test/analysis/plugins/BuildFile.xml:
```diff
@@ -0,0 +1,47 @@
+<library   name="PCCNTuplerPlugin" file="*.cc">
```
I think this file is not needed.
the code is already compiled by  RecoLuminosity/LumiProducer/test/BuildFile.xml

slava77 commented on RecoLuminosity/LumiProducer/test/BuildFile.xml:
```diff
@@ -1,5 +1,26 @@
+    <use   name="Alignment/OfflineValidation"/>
+    <use   name="CondFormats/SiPixelObjects"/>
+    <use   name="CondFormats/RunInfo"/>
+    <use   name="clhep"/>
```
Alignment/OfflineValidation and CondFormats/SiPixelObjects are not needed according to the libchecker

slava77 commented:
I think we can sign off after the need for  RecoLuminosity/LumiProducer/test/analysis/plugins/BuildFile.xml is clarified

capalmer85 committed:
remove old buildfile; clean new buildfile

capalmer85 commented:
Hi Slava,

Thank you for checking this.  Actually you were right on both counts.  The
extra buildfile was left by mistake and those two lines weren't needed for
the new buildfile.  I've pushed an update to the branch.

Cheers,
Chris

On Fri, Mar 6, 2015 at 6:18 PM, Slava Krutelyov <notifications@github.com>
wrote:

> I think we can sign off after the need for
> RecoLuminosity/LumiProducer/test/analysis/plugins/BuildFile.xml is clarified
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/6962#issuecomment-77654195>.
>

cmsbuild commented:
Pull request #6962 was updated. @cmsbuild, @cvuosalo, @nclopezo, @slava77 can you please check and sign again.

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/6962	8781d47
tested in CMSSW_5_3_26 /test area sign5326a/
the code compiles without issues and the test config provided is operational 
(after dropping ``` file:tempeos/cms ``` in the input file name)

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_5_3_X IBs unless changes or unless it breaks tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @smuzaffar, @ktf

cmsbuild commented:
+1
Tested at: 8781d47c73e8dadfd7fcfc5af7aef45ab956eb9e
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6962/3170/summary.html

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_5_3_X IBs unless changes (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @smuzaffar, @ktf

davidlange6 commented:
+1
